import grpc
from concurrent import futures
import greeter_pb2
import greeter_pb2_grpc
import threading
import sys
import time
import logging

logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s - %(levelname)s - %(message)s",
)

# In-memory user storage (for demonstration purposes only)
user_db = {'1@gmail.com': '1'}
db_lock = threading.Lock()

# List of server ports
server_ports = [50051, 50052, 50053]

class GreeterService(greeter_pb2_grpc.GreeterServicer):
    def GetIndexPage(self, request, context):
        return greeter_pb2.IndexResponse(html_content="<h1>Welcome to the Home Page</h1>")

    def Authenticate(self, request, context):
        username, password = request.username, request.password
        with db_lock:
            success = username in user_db and user_db[username] == password
        print(f"‚úÖ X√°c th·ª±c th√†nh c√¥ng cho {username}" if success else f"‚ùå X√°c th·ª±c th·∫•t b·∫°i cho {username}")
        return greeter_pb2.AuthResponse(success=success)

    def SyncUserData(self, request, context):
        username, password = request.username, request.password
        with db_lock:
            user_db[username] = password
        print(f"üì• ƒê·ªìng b·ªô t√†i kho·∫£n {username} t·ª´ server kh√°c")
        return greeter_pb2.SyncResponse(success=True)
    
    def Register(self, request, context):
        username, password = request.username, request.password
        logging.info(f"üì• Nh·∫≠n y√™u c·∫ßu ƒëƒÉng k√Ω t·ª´ {username}")

        # Ki·ªÉm tra t√†i kho·∫£n c√≥ trong DB c·ª•c b·ªô
        with db_lock:
            if username in user_db:
                logging.warning(f"‚ö†Ô∏è T√†i kho·∫£n {username} ƒë√£ t·ªìn t·∫°i trong server c·ª•c b·ªô.")
                return greeter_pb2.AuthResponse(success=False, message="T√†i kho·∫£n ƒë√£ t·ªìn t·∫°i.")

        # Ki·ªÉm tra t√†i kho·∫£n c√≥ trong c√°c server kh√°c
        for target_port in server_ports:
            if target_port == int(sys.argv[1]):  # B·ªè qua ch√≠nh server hi·ªán t·∫°i
                continue
            try:
                with grpc.insecure_channel(f'localhost:{target_port}') as channel:
                    stub = greeter_pb2_grpc.GreeterStub(channel)
                    check_request = greeter_pb2.CheckUserExistenceRequest(username=username)
                    response = stub.CheckUserExistence(check_request)
                    if response.exists:
                        logging.warning(f"‚ö†Ô∏è T√†i kho·∫£n {username} ƒë√£ t·ªìn t·∫°i tr√™n server kh√°c t·∫°i c·ªïng {target_port}")
                        return greeter_pb2.AuthResponse(success=False, message="T√†i kho·∫£n ƒë√£ t·ªìn t·∫°i tr√™n server kh√°c.")
            except grpc.RpcError as e:
                logging.error(f"‚ùå L·ªói khi ki·ªÉm tra server {target_port}: {e.details()}")

        # ƒêƒÉng k√Ω v√†o DB c·ª•c b·ªô
        with db_lock:
            user_db[username] = password
            logging.info(f"‚úÖ ƒêƒÉng k√Ω th√†nh c√¥ng t√†i kho·∫£n {username} trong server c·ª•c b·ªô.")

        # ƒê·ªìng b·ªô v·ªõi c√°c server trong Finger Table
        sync_with_finger_table(username, password, build_finger_table(int(sys.argv[1])))

        logging.info(f"‚úÖ Ho√†n th√†nh ƒëƒÉng k√Ω v√† ƒë·ªìng b·ªô t√†i kho·∫£n {username}")
        return greeter_pb2.AuthResponse(success=True, message="ƒêƒÉng k√Ω th√†nh c√¥ng.")


    def CheckUserExistence(self, request, context):
        with db_lock:
            exists = request.username in user_db
        return greeter_pb2.CheckUserExistenceResponse(exists=exists)

def build_finger_table(port):
    index = server_ports.index(port)
    finger_table = [server_ports[(index + 2 ** i) % len(server_ports)] for i in range(len(server_ports))]
    print(f"üìù Finger Table cho c·ªïng {port}: {finger_table}")
    return finger_table

def sync_with_finger_table(username, password, finger_table):
    for target_port in finger_table:
        target_server = f"localhost:{target_port}"
        try:
            with grpc.insecure_channel(target_server) as channel:
                grpc.channel_ready_future(channel).result(timeout=3)
                stub = greeter_pb2_grpc.GreeterStub(channel)
                request = greeter_pb2.SyncRequest(username=username, password=password)
                if stub.SyncUserData(request, timeout=3).success:
                    print(f"‚úÖ ƒê·ªìng b·ªô t√†i kho·∫£n {username} sang {target_server} th√†nh c√¥ng")
        except grpc.FutureTimeoutError:
            logging.warning(f"‚ö†Ô∏è Kh√¥ng th·ªÉ k·∫øt n·ªëi t·ªõi {target_server} (Timeout)")
        except grpc.RpcError as e:
            logging.error(f"‚ùå L·ªói ƒë·ªìng b·ªô v·ªõi {target_server}: {e.details()}")
            if "Connect" in str(e) and "refused" in str(e):
                logging.warning(f"‚ö†Ô∏è B·ªè qua server {target_server} v√¨ kh√¥ng th·ªÉ k·∫øt n·ªëi")

def periodic_sync(finger_table):
    print("üîÑ B·∫Øt ƒë·∫ßu ƒë·ªìng b·ªô ƒë·ªãnh k·ª≥...")
    with db_lock:
        for username, password in user_db.items():
            sync_with_finger_table(username, password, finger_table)
    threading.Timer(60, periodic_sync, [finger_table]).start()


def serve(port):
    server = grpc.server(futures.ThreadPoolExecutor(max_workers=50))
    greeter_pb2_grpc.add_GreeterServicer_to_server(GreeterService(), server)
    server.add_insecure_port(f'[::]:{port}')
    print(f"gRPC Server ƒëang ch·∫°y tr√™n c·ªïng {port}...")

    finger_table = build_finger_table(port)
    periodic_sync(finger_table)

    server.start()
    server.wait_for_termination()


if __name__ == "__main__":
    port = int(sys.argv[1]) if len(sys.argv) > 1 else 50051
    if port not in server_ports:
        print("‚ùå C·ªïng kh√¥ng h·ª£p l·ªá. Vui l√≤ng ch·ªçn t·ª´ 50051 ƒë·∫øn 50053.")
        sys.exit(1)
    serve(port)
